{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bde0bbf1",
   "metadata": {},
   "source": [
    "Extract Tables from PDF and convert to Excel sheet with Paddle OCR text detection and recognition"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cacff69d",
   "metadata": {},
   "source": [
    "#https://www.youtube.com/watch?v=HZh31OGiQRQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8b04546",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pdf2image in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (1.16.3)\n",
      "Requirement already satisfied: pillow in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pdf2image) (9.5.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: poppler-utils in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (0.1.0)\n",
      "Requirement already satisfied: Click>=7.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from poppler-utils) (8.1.3)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: paddlepaddle in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (2.5.0)\n",
      "Requirement already satisfied: httpx in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddlepaddle) (0.24.1)\n",
      "Requirement already satisfied: numpy>=1.13 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddlepaddle) (1.24.3)\n",
      "Requirement already satisfied: Pillow in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddlepaddle) (9.5.0)\n",
      "Requirement already satisfied: decorator in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddlepaddle) (5.1.1)\n",
      "Requirement already satisfied: astor in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddlepaddle) (0.8.1)\n",
      "Requirement already satisfied: paddle-bfloat==0.1.7 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddlepaddle) (0.1.7)\n",
      "Requirement already satisfied: opt-einsum==3.3.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddlepaddle) (3.3.0)\n",
      "Requirement already satisfied: protobuf>=3.20.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddlepaddle) (3.20.2)\n",
      "Requirement already satisfied: certifi in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from httpx->paddlepaddle) (2022.5.18.1)\n",
      "Requirement already satisfied: httpcore<0.18.0,>=0.15.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from httpx->paddlepaddle) (0.17.2)\n",
      "Requirement already satisfied: idna in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from httpx->paddlepaddle) (3.3)\n",
      "Requirement already satisfied: sniffio in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from httpx->paddlepaddle) (1.3.0)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from httpcore<0.18.0,>=0.15.0->httpx->paddlepaddle) (0.14.0)\n",
      "Requirement already satisfied: anyio<5.0,>=3.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from httpcore<0.18.0,>=0.15.0->httpx->paddlepaddle) (3.7.0)\n",
      "Requirement already satisfied: exceptiongroup in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from anyio<5.0,>=3.0->httpcore<0.18.0,>=0.15.0->httpx->paddlepaddle) (1.1.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: layoutparser in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (0.3.4)\n",
      "Requirement already satisfied: numpy in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from layoutparser) (1.24.3)\n",
      "Requirement already satisfied: opencv-python in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from layoutparser) (4.6.0.66)\n",
      "Requirement already satisfied: scipy in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from layoutparser) (1.11.0)\n",
      "Requirement already satisfied: pandas in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from layoutparser) (2.0.2)\n",
      "Requirement already satisfied: pillow in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from layoutparser) (9.5.0)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from layoutparser) (6.0)\n",
      "Requirement already satisfied: iopath in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from layoutparser) (0.1.10)\n",
      "Requirement already satisfied: pdfplumber in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from layoutparser) (0.10.1)\n",
      "Requirement already satisfied: pdf2image in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from layoutparser) (1.16.3)\n",
      "Requirement already satisfied: tqdm in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from iopath->layoutparser) (4.65.0)\n",
      "Requirement already satisfied: typing-extensions in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from iopath->layoutparser) (4.5.0)\n",
      "Requirement already satisfied: portalocker in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from iopath->layoutparser) (2.7.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pandas->layoutparser) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pandas->layoutparser) (2023.3)\n",
      "Requirement already satisfied: tzdata>=2022.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pandas->layoutparser) (2023.3)\n",
      "Requirement already satisfied: pdfminer.six==20221105 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pdfplumber->layoutparser) (20221105)\n",
      "Requirement already satisfied: pypdfium2>=4.18.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pdfplumber->layoutparser) (4.18.0)\n",
      "Requirement already satisfied: charset-normalizer>=2.0.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pdfminer.six==20221105->pdfplumber->layoutparser) (2.0.12)\n",
      "Requirement already satisfied: cryptography>=36.0.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pdfminer.six==20221105->pdfplumber->layoutparser) (41.0.1)\n",
      "Requirement already satisfied: six>=1.5 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from python-dateutil>=2.8.2->pandas->layoutparser) (1.16.0)\n",
      "Requirement already satisfied: cffi>=1.12 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from cryptography>=36.0.0->pdfminer.six==20221105->pdfplumber->layoutparser) (1.15.1)\n",
      "Requirement already satisfied: pycparser in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from cffi>=1.12->cryptography>=36.0.0->pdfminer.six==20221105->pdfplumber->layoutparser) (2.21)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Collecting paddle\n",
      "  Using cached paddle-1.0.2.tar.gz (579 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpython setup.py egg_info\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[8 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"<string>\", line 2, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"/private/var/folders/m5/8mkz0vqn4v5dmsqp6w27mb880000gn/T/pip-install-covkd1rf/paddle_1dbb16f717ad41f79642c3370275ae03/setup.py\", line 3, in <module>\n",
      "  \u001b[31m   \u001b[0m     import paddle\n",
      "  \u001b[31m   \u001b[0m   File \"/private/var/folders/m5/8mkz0vqn4v5dmsqp6w27mb880000gn/T/pip-install-covkd1rf/paddle_1dbb16f717ad41f79642c3370275ae03/paddle/__init__.py\", line 5, in <module>\n",
      "  \u001b[31m   \u001b[0m     import common, dual, tight, data, prox\n",
      "  \u001b[31m   \u001b[0m ModuleNotFoundError: No module named 'common'\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[?25h\u001b[1;31merror\u001b[0m: \u001b[1mmetadata-generation-failed\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m Encountered error while generating package metadata.\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This is an issue with the package mentioned above, not pip.\n",
      "\u001b[1;36mhint\u001b[0m: See above for details.\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: tensorflow in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (2.13.0)\n",
      "Requirement already satisfied: tensorflow-macos==2.13.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow) (2.13.0)\n",
      "Requirement already satisfied: absl-py>=1.0.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (1.4.0)\n",
      "Requirement already satisfied: astunparse>=1.6.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (1.6.3)\n",
      "Requirement already satisfied: flatbuffers>=23.1.21 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (23.5.26)\n",
      "Requirement already satisfied: gast<=0.4.0,>=0.2.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (0.4.0)\n",
      "Requirement already satisfied: google-pasta>=0.1.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (0.2.0)\n",
      "Requirement already satisfied: h5py>=2.9.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (3.9.0)\n",
      "Requirement already satisfied: libclang>=13.0.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (16.0.0)\n",
      "Requirement already satisfied: numpy<=1.24.3,>=1.22 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (1.24.3)\n",
      "Requirement already satisfied: opt-einsum>=2.3.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (3.3.0)\n",
      "Requirement already satisfied: packaging in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (23.1)\n",
      "Collecting protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3 (from tensorflow-macos==2.13.0->tensorflow)\n",
      "  Obtaining dependency information for protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3 from https://files.pythonhosted.org/packages/cb/d3/a164038605494d49acc4f9cda1c0bc200b96382c53edd561387263bb181d/protobuf-4.23.4-cp37-abi3-macosx_10_9_universal2.whl.metadata\n",
      "  Using cached protobuf-4.23.4-cp37-abi3-macosx_10_9_universal2.whl.metadata (540 bytes)\n",
      "Requirement already satisfied: setuptools in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (68.0.0)\n",
      "Requirement already satisfied: six>=1.12.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (1.16.0)\n",
      "Requirement already satisfied: termcolor>=1.1.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (2.3.0)\n",
      "Requirement already satisfied: typing-extensions<4.6.0,>=3.6.6 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (4.5.0)\n",
      "Requirement already satisfied: wrapt>=1.11.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (1.15.0)\n",
      "Requirement already satisfied: grpcio<2.0,>=1.24.3 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (1.56.0)\n",
      "Requirement already satisfied: tensorboard<2.14,>=2.13 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (2.13.0)\n",
      "Requirement already satisfied: tensorflow-estimator<2.14,>=2.13.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (2.13.0)\n",
      "Requirement already satisfied: keras<2.14,>=2.13.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow-macos==2.13.0->tensorflow) (2.13.1)\n",
      "Requirement already satisfied: wheel<1.0,>=0.23.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from astunparse>=1.6.0->tensorflow-macos==2.13.0->tensorflow) (0.40.0)\n",
      "Requirement already satisfied: google-auth<3,>=1.6.3 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (2.22.0)\n",
      "Requirement already satisfied: google-auth-oauthlib<1.1,>=0.5 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (1.0.0)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (3.4.3)\n",
      "Requirement already satisfied: requests<3,>=2.21.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (2.27.1)\n",
      "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (0.7.1)\n",
      "Requirement already satisfied: werkzeug>=1.0.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (2.3.6)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (5.3.1)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (0.3.0)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (4.9)\n",
      "Requirement already satisfied: urllib3<2.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (1.26.9)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from google-auth-oauthlib<1.1,>=0.5->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (1.3.1)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests<3,>=2.21.0->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (2022.5.18.1)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests<3,>=2.21.0->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (2.0.12)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests<3,>=2.21.0->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (3.3)\n",
      "Requirement already satisfied: MarkupSafe>=2.1.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from werkzeug>=1.0.1->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (2.1.3)\n",
      "Requirement already satisfied: pyasn1<0.6.0,>=0.4.6 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (0.5.0)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<1.1,>=0.5->tensorboard<2.14,>=2.13->tensorflow-macos==2.13.0->tensorflow) (3.2.2)\n",
      "Using cached protobuf-4.23.4-cp37-abi3-macosx_10_9_universal2.whl (400 kB)\n",
      "Installing collected packages: protobuf\n",
      "  Attempting uninstall: protobuf\n",
      "    Found existing installation: protobuf 3.20.2\n",
      "    Uninstalling protobuf-3.20.2:\n",
      "      Successfully uninstalled protobuf-3.20.2\n",
      "Successfully installed protobuf-4.23.4\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install pdf2image\n",
    "%pip install poppler-utils\n",
    "%pip install paddlepaddle\n",
    "%pip install layoutparser\n",
    "%pip install paddle\n",
    "%pip install tensorflow"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58e118ae",
   "metadata": {},
   "source": [
    "# Convert pdf to image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f3ec240f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pdf2image import convert_from_path\n",
    "images = convert_from_path('./report.pdf', poppler_path='/opt/homebrew/bin')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a3e0dbd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: pages: File exists\n"
     ]
    }
   ],
   "source": [
    "%mkdir pages\n",
    "\n",
    "for i in range(len(images)):\n",
    "    images[i].save('pages/page'+str(i)+'.jpg','JPEG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c927594a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PublayNet\n",
    "\n",
    "#PaddlePaddle/PaddleOCR Github, layout detection\n",
    "#ppstructure/layout/ "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e8f70f7",
   "metadata": {},
   "source": [
    "# Layout extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7c543511",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cv2\n",
    "import layoutparser as lp\n",
    "import paddle\n",
    "\n",
    "image = cv2.imread(\"./pages/page0.jpg\")\n",
    "image = image[..., ::-1]\n",
    "device = paddle.set_device('cpu')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "adef5589",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# load model\n",
    "# model = lp.PaddleDetectionLayoutModel(config_path=\"lp://PubLayNet/ppyolov2_r50vd_dcn_365e_publaynet/config\",\n",
    "#                                 threshold=0.5,\n",
    "#                                 label_map={0: \"Text\", 1: \"Title\", 2: \"List\", 3:\"Table\", 4:\"Figure\"},\n",
    "#                                 enforce_cpu=False,\n",
    "#                                 enable_mkldnn=True)\n",
    "\n",
    "# ORIGINAL config_path = \"lp://PubLayNet/ppyolov2_r50vd_dcn_365e_publaynet/config\"\n",
    "\n",
    "model = lp.PaddleDetectionLayoutModel(config_path=\"lp://paddledetection/TableBank/ppyolov2_r50vd_dcn_365e\",\n",
    "                                label_map={0: \"Text\", 1: \"Title\", 2: \"List\", 3:\"Table\", 4:\"Figure\"},\n",
    "                                enforce_cpu=True) \n",
    "\n",
    "\n",
    "# def init_model():\n",
    "#     model = lp.Detectron2LayoutModel(\n",
    "#         os.path.join(os.getcwd(), 'code/ParseServer/models/PubLayNet/faster_rcnn_R_50_FPN_3x/config.yml'),\n",
    "#         os.path.join(os.getcwd(),\n",
    "#                      'code/ParseServer/models/PubLayNet/faster_rcnn_R_50_FPN_3x/model_final.pth'),\n",
    "#         {0: \"Text\", 1: \"Title\", 2: \"List\", 3: \"Table\", 4: \"Figure\"})\n",
    "#     return model\n",
    "\n",
    "\n",
    "# model = init_model()\n",
    "\n",
    "\n",
    "#detect\n",
    "layout = model.detect(image)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6ac66e36",
   "metadata": {},
   "outputs": [],
   "source": [
    "from layoutparser import PaddleDetectionLayoutModel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "345bda0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_only_effdet_model():\n",
    "\n",
    "    # When all the backeds are not installed, it should \n",
    "    # elicit only ImportErrors\n",
    "    \n",
    "    config = \"lp://PubLayNet/faster_rcnn_R_50_FPN_3x/config\"\n",
    "    model = Detectron2LayoutModel(config)\n",
    "    image = cv2.imread(\"./pages/page0.jpg\")\n",
    "    layout = model.detect(image)\n",
    "    \n",
    "    with pytest.raises(ImportError):\n",
    "        from layoutparser import EfficientDetLayoutModel\n",
    "        from layoutparser import PaddleDetectionLayoutModel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ec580c8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pip install -U layoutparser-0.0.1-py3-none-any.whl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "365750c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: paddleocr>=2.0.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (2.6.1.3)\n",
      "Requirement already satisfied: shapely in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (2.0.1)\n",
      "Requirement already satisfied: scikit-image in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (0.21.0)\n",
      "Requirement already satisfied: imgaug in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (0.4.0)\n",
      "Requirement already satisfied: pyclipper in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (1.3.0.post4)\n",
      "Requirement already satisfied: lmdb in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (1.4.1)\n",
      "Requirement already satisfied: tqdm in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (4.65.0)\n",
      "Requirement already satisfied: numpy in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (1.24.3)\n",
      "Requirement already satisfied: visualdl in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (2.5.3)\n",
      "Requirement already satisfied: rapidfuzz in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (3.1.1)\n",
      "Requirement already satisfied: opencv-python<=4.6.0.66 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (4.6.0.66)\n",
      "Requirement already satisfied: opencv-contrib-python<=4.6.0.66 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (4.6.0.66)\n",
      "Requirement already satisfied: cython in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (0.29.35)\n",
      "Requirement already satisfied: lxml in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (4.9.2)\n",
      "Requirement already satisfied: premailer in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (3.10.0)\n",
      "Requirement already satisfied: openpyxl in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (3.1.2)\n",
      "Requirement already satisfied: attrdict in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (2.0.1)\n",
      "Requirement already satisfied: PyMuPDF<1.21.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (1.20.2)\n",
      "Requirement already satisfied: python-docx in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (0.8.11)\n",
      "Requirement already satisfied: beautifulsoup4 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (4.12.2)\n",
      "Requirement already satisfied: fonttools>=4.24.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (4.40.0)\n",
      "Requirement already satisfied: fire>=0.3.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (0.5.0)\n",
      "Requirement already satisfied: pdf2docx in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from paddleocr>=2.0.1) (0.5.6)\n",
      "Requirement already satisfied: six in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from fire>=0.3.0->paddleocr>=2.0.1) (1.16.0)\n",
      "Requirement already satisfied: termcolor in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from fire>=0.3.0->paddleocr>=2.0.1) (2.3.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from beautifulsoup4->paddleocr>=2.0.1) (2.4.1)\n",
      "Requirement already satisfied: scipy in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from imgaug->paddleocr>=2.0.1) (1.11.0)\n",
      "Requirement already satisfied: Pillow in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from imgaug->paddleocr>=2.0.1) (9.5.0)\n",
      "Requirement already satisfied: matplotlib in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from imgaug->paddleocr>=2.0.1) (3.7.1)\n",
      "Requirement already satisfied: imageio in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from imgaug->paddleocr>=2.0.1) (2.31.1)\n",
      "Requirement already satisfied: networkx>=2.8 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from scikit-image->paddleocr>=2.0.1) (3.1)\n",
      "Requirement already satisfied: tifffile>=2022.8.12 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from scikit-image->paddleocr>=2.0.1) (2023.4.12)\n",
      "Requirement already satisfied: PyWavelets>=1.1.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from scikit-image->paddleocr>=2.0.1) (1.4.1)\n",
      "Requirement already satisfied: packaging>=21 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from scikit-image->paddleocr>=2.0.1) (23.1)\n",
      "Requirement already satisfied: lazy_loader>=0.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from scikit-image->paddleocr>=2.0.1) (0.2)\n",
      "Requirement already satisfied: et-xmlfile in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from openpyxl->paddleocr>=2.0.1) (1.1.0)\n",
      "Requirement already satisfied: cssselect in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from premailer->paddleocr>=2.0.1) (1.2.0)\n",
      "Requirement already satisfied: cssutils in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from premailer->paddleocr>=2.0.1) (2.7.1)\n",
      "Requirement already satisfied: requests in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from premailer->paddleocr>=2.0.1) (2.27.1)\n",
      "Requirement already satisfied: cachetools in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from premailer->paddleocr>=2.0.1) (5.3.1)\n",
      "Requirement already satisfied: bce-python-sdk in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from visualdl->paddleocr>=2.0.1) (0.8.83)\n",
      "Requirement already satisfied: flask>=1.1.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from visualdl->paddleocr>=2.0.1) (2.3.2)\n",
      "Requirement already satisfied: Flask-Babel>=3.0.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from visualdl->paddleocr>=2.0.1) (3.1.0)\n",
      "Requirement already satisfied: protobuf>=3.20.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from visualdl->paddleocr>=2.0.1) (4.23.4)\n",
      "Requirement already satisfied: pandas in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from visualdl->paddleocr>=2.0.1) (2.0.2)\n",
      "Requirement already satisfied: rarfile in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from visualdl->paddleocr>=2.0.1) (4.0)\n",
      "Requirement already satisfied: psutil in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from visualdl->paddleocr>=2.0.1) (5.9.5)\n",
      "Requirement already satisfied: Werkzeug>=2.3.3 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from flask>=1.1.1->visualdl->paddleocr>=2.0.1) (2.3.6)\n",
      "Requirement already satisfied: Jinja2>=3.1.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from flask>=1.1.1->visualdl->paddleocr>=2.0.1) (3.1.2)\n",
      "Requirement already satisfied: itsdangerous>=2.1.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from flask>=1.1.1->visualdl->paddleocr>=2.0.1) (2.1.2)\n",
      "Requirement already satisfied: click>=8.1.3 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from flask>=1.1.1->visualdl->paddleocr>=2.0.1) (8.1.3)\n",
      "Requirement already satisfied: blinker>=1.6.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from flask>=1.1.1->visualdl->paddleocr>=2.0.1) (1.6.2)\n",
      "Requirement already satisfied: Babel>=2.12 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from Flask-Babel>=3.0.0->visualdl->paddleocr>=2.0.1) (2.12.1)\n",
      "Requirement already satisfied: pytz>=2022.7 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from Flask-Babel>=3.0.0->visualdl->paddleocr>=2.0.1) (2023.3)\n",
      "Requirement already satisfied: pycryptodome>=3.8.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from bce-python-sdk->visualdl->paddleocr>=2.0.1) (3.18.0)\n",
      "Requirement already satisfied: future>=0.6.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from bce-python-sdk->visualdl->paddleocr>=2.0.1) (0.18.3)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from matplotlib->imgaug->paddleocr>=2.0.1) (1.1.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from matplotlib->imgaug->paddleocr>=2.0.1) (0.11.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from matplotlib->imgaug->paddleocr>=2.0.1) (1.4.4)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from matplotlib->imgaug->paddleocr>=2.0.1) (3.1.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from matplotlib->imgaug->paddleocr>=2.0.1) (2.8.2)\n",
      "Requirement already satisfied: tzdata>=2022.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from pandas->visualdl->paddleocr>=2.0.1) (2023.3)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests->premailer->paddleocr>=2.0.1) (1.26.9)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests->premailer->paddleocr>=2.0.1) (2022.5.18.1)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests->premailer->paddleocr>=2.0.1) (2.0.12)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from requests->premailer->paddleocr>=2.0.1) (3.3)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from Jinja2>=3.1.2->flask>=1.1.1->visualdl->paddleocr>=2.0.1) (2.1.3)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Collecting protobuf==3.20.2\n",
      "  Using cached protobuf-3.20.2-py2.py3-none-any.whl (162 kB)\n",
      "Installing collected packages: protobuf\n",
      "  Attempting uninstall: protobuf\n",
      "    Found existing installation: protobuf 4.23.4\n",
      "    Uninstalling protobuf-4.23.4:\n",
      "      Successfully uninstalled protobuf-4.23.4\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "tensorflow-macos 2.13.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3, but you have protobuf 3.20.2 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed protobuf-3.20.2\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Line magic function `%git` not found.\n"
     ]
    }
   ],
   "source": [
    "# %pip install paddlepaddle-gpu\n",
    "%pip install \"paddleocr>=2.0.1\"\n",
    "%pip install protobuf==3.20.2\n",
    "%git clone https://github.com/PaddlePaddle/PaddleOCR.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1b7334f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1048, 1680)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# paddleOCR and drawocr\n",
    "from paddleocr import PaddleOCR, draw_ocr\n",
    "\n",
    "# extra utilities\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "# image = cv2.imread('./report_copy_removed.pdf')\n",
    "test_image = cv2.imread('./ppocr_img/imgs_en/img_12.jpg')\n",
    "len(test_image), len(test_image[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "96088135",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2023/07/24 16:57:55] ppocr DEBUG: Namespace(help='==SUPPRESS==', use_gpu=False, use_xpu=False, use_npu=False, ir_optim=True, use_tensorrt=False, min_subgraph_size=15, precision='fp32', gpu_mem=500, image_dir=None, page_num=0, det_algorithm='DB', det_model_dir='/Users/winstontang/.paddleocr/whl/det/en/en_PP-OCRv3_det_infer', det_limit_side_len=960, det_limit_type='max', det_box_type='quad', det_db_thresh=0.3, det_db_box_thresh=0.6, det_db_unclip_ratio=1.5, max_batch_size=10, use_dilation=False, det_db_score_mode='fast', det_east_score_thresh=0.8, det_east_cover_thresh=0.1, det_east_nms_thresh=0.2, det_sast_score_thresh=0.5, det_sast_nms_thresh=0.2, det_pse_thresh=0, det_pse_box_thresh=0.85, det_pse_min_area=16, det_pse_scale=1, scales=[8, 16, 32], alpha=1.0, beta=1.0, fourier_degree=5, rec_algorithm='SVTR_LCNet', rec_model_dir='/Users/winstontang/.paddleocr/whl/rec/en/en_PP-OCRv3_rec_infer', rec_image_inverse=True, rec_image_shape='3, 48, 320', rec_batch_num=6, max_text_length=25, rec_char_dict_path='/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/paddleocr/ppocr/utils/en_dict.txt', use_space_char=True, vis_font_path='./doc/fonts/simfang.ttf', drop_score=0.5, e2e_algorithm='PGNet', e2e_model_dir=None, e2e_limit_side_len=768, e2e_limit_type='max', e2e_pgnet_score_thresh=0.5, e2e_char_dict_path='./ppocr/utils/ic15_dict.txt', e2e_pgnet_valid_set='totaltext', e2e_pgnet_mode='fast', use_angle_cls=True, cls_model_dir='/Users/winstontang/.paddleocr/whl/cls/ch_ppocr_mobile_v2.0_cls_infer', cls_image_shape='3, 48, 192', label_list=['0', '180'], cls_batch_num=6, cls_thresh=0.9, enable_mkldnn=False, cpu_threads=10, use_pdserving=False, warmup=False, sr_model_dir=None, sr_image_shape='3, 32, 128', sr_batch_num=1, draw_img_save_dir='./inference_results', save_crop_res=False, crop_res_save_dir='./output', use_mp=False, total_process_num=1, process_id=0, benchmark=False, save_log_path='./log_output/', show_log=True, use_onnx=False, output='./output', table_max_len=488, table_algorithm='TableAttn', table_model_dir=None, merge_no_span_structure=True, table_char_dict_path=None, layout_model_dir=None, layout_dict_path=None, layout_score_threshold=0.5, layout_nms_threshold=0.5, kie_algorithm='LayoutXLM', ser_model_dir=None, re_model_dir=None, use_visual_backbone=True, ser_dict_path='../train_data/XFUND/class_list_xfun.txt', ocr_order_method=None, mode='structure', image_orientation=False, layout=True, table=True, ocr=True, recovery=False, use_pdf2docx_api=False, lang='en', det=True, rec=True, type='ocr', ocr_version='PP-OCRv3', structure_version='PP-StructureV2')\n",
      "[2023/07/24 16:57:56] ppocr DEBUG: dt_boxes num : 11, elapse : 0.3780059814453125\n",
      "[2023/07/24 16:57:56] ppocr DEBUG: cls num  : 11, elapse : 0.09166359901428223\n",
      "[2023/07/24 16:57:58] ppocr DEBUG: rec_res num  : 11, elapse : 2.1723477840423584\n",
      "[[[441.0, 174.0], [1166.0, 176.0], [1165.0, 222.0], [441.0, 221.0]], ('ACKNOWLEDGEMENTS', 0.9971134662628174)]\n",
      "[[[403.0, 346.0], [1204.0, 348.0], [1204.0, 384.0], [402.0, 383.0]], ('We would like to thank all the designers and', 0.9761400818824768)]\n",
      "[[[403.0, 396.0], [1204.0, 398.0], [1204.0, 434.0], [402.0, 433.0]], ('contributors who have been involved in the', 0.9791957139968872)]\n",
      "[[[399.0, 446.0], [1207.0, 443.0], [1208.0, 484.0], [399.0, 488.0]], ('production of this book; their contributions', 0.9889592528343201)]\n",
      "[[[401.0, 500.0], [1208.0, 500.0], [1208.0, 534.0], [401.0, 534.0]], ('have been indispensable to its creation. We', 0.9555092453956604)]\n",
      "[[[399.0, 550.0], [1209.0, 548.0], [1209.0, 583.0], [399.0, 584.0]], ('would also like to express our gratitude to all', 0.9905261397361755)]\n",
      "[[[399.0, 600.0], [1207.0, 598.0], [1208.0, 634.0], [399.0, 636.0]], ('the producers for their invaluable opinions', 0.9781714081764221)]\n",
      "[[[399.0, 648.0], [1207.0, 646.0], [1208.0, 686.0], [399.0, 688.0]], ('and assistance throughout this project. And to', 0.9883645176887512)]\n",
      "[[[399.0, 702.0], [1209.0, 698.0], [1209.0, 734.0], [399.0, 738.0]], ('the many others whose names are not credited', 0.9579494595527649)]\n",
      "[[[399.0, 750.0], [1211.0, 750.0], [1211.0, 789.0], [399.0, 789.0]], ('but have made specific input in this book, we', 0.9702869057655334)]\n",
      "[[[397.0, 802.0], [1090.0, 800.0], [1090.0, 839.0], [397.0, 841.0]], ('thank you for your continuous support.', 0.9978144764900208)]\n"
     ]
    }
   ],
   "source": [
    "image_path = './ilovepdf_pages-to-jpg/report_copy_removed_page-0001.jpg'\n",
    "test_image_path = './ppocr_img/imgs_en/img_12.jpg'\n",
    "\n",
    "ocr = PaddleOCR(use_angle_cls=True, lang='en')\n",
    "\n",
    "test_result = ocr.ocr(test_image_path, cls=True)\n",
    "for idx in range(len(test_result)):\n",
    "    res = test_result[idx]\n",
    "    for line in res:\n",
    "        print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2a7aeadf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2023/07/24 16:57:59] ppocr DEBUG: dt_boxes num : 86, elapse : 0.3730309009552002\n",
      "[2023/07/24 16:57:59] ppocr DEBUG: cls num  : 86, elapse : 0.4962480068206787\n",
      "[2023/07/24 16:58:05] ppocr DEBUG: rec_res num  : 86, elapse : 5.801584005355835\n",
      "[[[308.0, 77.0], [544.0, 77.0], [544.0, 100.0], [308.0, 100.0]], ('Accession ID: NGSO-221209001', 0.9855970740318298)]\n",
      "[[[1105.0, 79.0], [1199.0, 79.0], [1199.0, 98.0], [1105.0, 98.0]], ('Page 5 of 76', 0.9598619341850281)]\n",
      "[[[300.0, 96.0], [591.0, 96.0], [591.0, 119.0], [300.0, 119.0]], ('Patient Name: LAN ZHENG,MEI-CHAN', 0.96770840883255)]\n",
      "[[[329.0, 117.0], [525.0, 117.0], [525.0, 139.0], [329.0, 139.0]], ('Diagnosis: gastric cancer', 0.9960033893585205)]\n",
      "[[[85.0, 143.0], [172.0, 143.0], [172.0, 162.0], [85.0, 162.0]], ('YOUR LAB', 0.9481184482574463)]\n",
      "[[[310.0, 134.0], [511.0, 134.0], [511.0, 156.0], [310.0, 156.0]], ('Report Date: Dec 19,2022', 0.9939796328544617)]\n",
      "[[[700.0, 220.0], [899.0, 220.0], [899.0, 242.0], [700.0, 242.0]], ('Approved Therapies', 0.9706971049308777)]\n",
      "[[[883.0, 275.0], [1045.0, 275.0], [1045.0, 297.0], [883.0, 297.0]], ('Associated With', 0.9997997879981995)]\n",
      "[[[1076.0, 273.0], [1152.0, 273.0], [1152.0, 297.0], [1076.0, 297.0]], ('Clinical', 0.9997788071632385)]\n",
      "[[[78.0, 299.0], [218.0, 299.0], [218.0, 321.0], [78.0, 321.0]], ('Gene /Variant', 0.981497049331665)]\n",
      "[[[288.0, 301.0], [437.0, 301.0], [437.0, 323.0], [288.0, 323.0]], ('Allelic Fraction', 0.9724547266960144)]\n",
      "[[[497.0, 301.0], [643.0, 301.0], [643.0, 323.0], [497.0, 323.0]], ('Gastric Cancer', 0.9985162019729614)]\n",
      "[[[693.0, 302.0], [863.0, 302.0], [863.0, 320.0], [693.0, 320.0]], ('Other Indications', 0.9784754514694214)]\n",
      "[[[883.0, 302.0], [991.0, 302.0], [991.0, 321.0], [883.0, 321.0]], ('Resistance', 0.9960113763809204)]\n",
      "[[[1074.0, 299.0], [1133.0, 299.0], [1133.0, 323.0], [1074.0, 323.0]], ('Trials', 0.9989027976989746)]\n",
      "[[[691.0, 349.0], [811.0, 349.0], [811.0, 368.0], [691.0, 368.0]], ('pertuzumab', 0.9986637234687805)]\n",
      "[[[689.0, 383.0], [816.0, 383.0], [816.0, 402.0], [689.0, 402.0]], ('trastuzumab', 0.9996299743652344)]\n",
      "[[[691.0, 409.0], [793.0, 409.0], [793.0, 428.0], [691.0, 428.0]], ('emtansine', 0.9990456700325012)]\n",
      "[[[689.0, 443.0], [778.0, 443.0], [778.0, 462.0], [689.0, 462.0]], ('tucatinib', 0.9996497631072998)]\n",
      "[[[76.0, 483.0], [172.0, 483.0], [172.0, 512.0], [76.0, 512.0]], ('ERBB3', 0.99937504529953)]\n",
      "[[[289.0, 493.0], [414.0, 493.0], [414.0, 512.0], [289.0, 512.0]], ('48.0% (of 629', 0.9947896599769592)]\n",
      "[[[691.0, 493.0], [766.0, 493.0], [766.0, 512.0], [691.0, 512.0]], ('afatinib', 0.9984056949615479)]\n",
      "[[[74.0, 516.0], [181.0, 512.0], [182.0, 536.0], [75.0, 540.0]], ('c.3380G>A', 0.99739009141922)]\n",
      "[[[286.0, 517.0], [348.0, 517.0], [348.0, 541.0], [286.0, 541.0]], ('reads)', 0.9992241263389587)]\n",
      "[[[688.0, 526.0], [785.0, 526.0], [785.0, 550.0], [688.0, 550.0]], ('bosutinib', 0.9999136328697205)]\n",
      "[[[74.0, 542.0], [162.0, 538.0], [163.0, 562.0], [75.0, 566.0]], ('p.Rl127H', 0.886833906173706)]\n",
      "[[[76.0, 567.0], [218.0, 567.0], [218.0, 590.0], [76.0, 590.0]], ('g.56101239G>A', 0.9894739985466003)]\n",
      "[[[691.0, 564.0], [807.0, 564.0], [807.0, 583.0], [691.0, 583.0]], ('dacomitinib', 0.9990218877792358)]\n",
      "[[[78.0, 593.0], [133.0, 593.0], [133.0, 612.0], [78.0, 612.0]], ('Tier 3', 0.9839785695075989)]\n",
      "[[[691.0, 600.0], [780.0, 600.0], [780.0, 619.0], [691.0, 619.0]], ('neratinib', 0.9990379810333252)]\n",
      "[[[76.0, 617.0], [178.0, 617.0], [178.0, 641.0], [76.0, 641.0]], ('Uncertain', 0.9999268054962158)]\n",
      "[[[689.0, 632.0], [804.0, 632.0], [804.0, 657.0], [689.0, 657.0]], ('osimertinib', 0.9998282790184021)]\n",
      "[[[78.0, 645.0], [199.0, 645.0], [199.0, 663.0], [78.0, 663.0]], ('Significance', 0.9960393309593201)]\n",
      "[[[76.0, 689.0], [168.0, 689.0], [168.0, 713.0], [76.0, 713.0]], ('FGFR4', 0.9996960759162903)]\n",
      "[[[288.0, 693.0], [475.0, 693.0], [475.0, 715.0], [288.0, 715.0]], ('48.0% (of 145 reads)', 0.9888114929199219)]\n",
      "[[[691.0, 694.0], [790.0, 694.0], [790.0, 713.0], [691.0, 713.0]], ('erdafitinib', 0.9968708157539368)]\n",
      "[[[78.0, 718.0], [168.0, 718.0], [168.0, 737.0], [78.0, 737.0]], ('c.1871A>C', 0.9381946921348572)]\n",
      "[[[691.0, 729.0], [792.0, 729.0], [792.0, 748.0], [691.0, 748.0]], ('futibatinib', 0.9984834790229797)]\n",
      "[[[74.0, 743.0], [159.0, 739.0], [160.0, 763.0], [75.0, 767.0]], ('p.N624T', 0.9999807476997375)]\n",
      "[[[76.0, 767.0], [228.0, 765.0], [229.0, 789.0], [76.0, 791.0]], ('g.177096106A>C', 0.9959034323692322)]\n",
      "[[[691.0, 765.0], [797.0, 765.0], [797.0, 784.0], [691.0, 784.0]], ('infigratinib', 0.9973002076148987)]\n",
      "[[[76.0, 788.0], [137.0, 793.0], [136.0, 817.0], [74.0, 812.0]], ('Tier 3', 0.9933111071586609)]\n",
      "[[[689.0, 798.0], [788.0, 798.0], [788.0, 822.0], [689.0, 822.0]], ('lenvatinib', 0.9940023422241211)]\n",
      "[[[75.0, 814.0], [179.0, 818.0], [178.0, 843.0], [74.0, 838.0]], ('Uncertain', 0.999927282333374)]\n",
      "[[[76.0, 842.0], [201.0, 842.0], [201.0, 866.0], [76.0, 866.0]], ('Significance', 0.9996901154518127)]\n",
      "[[[687.0, 834.0], [786.0, 830.0], [787.0, 854.0], [688.0, 858.0]], ('ponatinib', 0.9998382925987244)]\n",
      "[[[691.0, 871.0], [804.0, 871.0], [804.0, 890.0], [691.0, 890.0]], ('regorafenib', 0.9978370666503906)]\n",
      "[[[76.0, 914.0], [149.0, 914.0], [149.0, 940.0], [76.0, 940.0]], ('JAK3', 0.9996762275695801)]\n",
      "[[[288.0, 920.0], [411.0, 920.0], [411.0, 938.0], [288.0, 938.0]], ('58.0% (of 262', 0.9797366261482239)]\n",
      "[[[688.0, 914.0], [790.0, 918.0], [789.0, 942.0], [687.0, 938.0]], ('ruxolitinib', 0.9999261498451233)]\n",
      "[[[76.0, 945.0], [180.0, 945.0], [180.0, 964.0], [76.0, 964.0]], ('c.2660G>A', 0.989569902420044)]\n",
      "[[[286.0, 944.0], [350.0, 944.0], [350.0, 968.0], [286.0, 968.0]], ('reads)', 0.9996784329414368)]\n",
      "[[[78.0, 971.0], [161.0, 971.0], [161.0, 990.0], [78.0, 990.0]], ('p.R887H', 0.9998840689659119)]\n",
      "[[[74.0, 994.0], [216.0, 988.0], [217.0, 1012.0], [75.0, 1018.0]], ('g.17832539C>T', 0.9960629343986511)]\n",
      "[[[76.0, 1015.0], [136.0, 1020.0], [134.0, 1044.0], [74.0, 1039.0]], ('Tier 3', 0.9973652362823486)]\n",
      "[[[75.0, 1041.0], [179.0, 1045.0], [178.0, 1069.0], [74.0, 1065.0]], ('Uncertain', 0.9999268054962158)]\n",
      "[[[76.0, 1069.0], [201.0, 1069.0], [201.0, 1093.0], [76.0, 1093.0]], ('Significance', 0.9996614456176758)]\n",
      "[[[80.0, 1131.0], [693.0, 1131.0], [693.0, 1155.0], [80.0, 1155.0]], ('Variants of Unknown Clinical Significance', 0.9925547242164612)]\n",
      "[[[75.0, 1176.0], [220.0, 1177.0], [220.0, 1202.0], [74.0, 1200.0]], ('Gene / Variant', 0.9697583913803101)]\n",
      "[[[283.0, 1176.0], [435.0, 1177.0], [435.0, 1202.0], [282.0, 1200.0]], (' Allelic Fraction', 0.9735027551651001)]\n",
      "[[[494.0, 1177.0], [587.0, 1177.0], [587.0, 1201.0], [494.0, 1201.0]], ('Function', 0.9999287128448486)]\n",
      "[[[686.0, 1176.0], [821.0, 1177.0], [821.0, 1202.0], [686.0, 1200.0]], ('Classification', 0.9998522996902466)]\n",
      "[[[882.0, 1181.0], [1003.0, 1181.0], [1003.0, 1200.0], [882.0, 1200.0]], ('Assessment', 0.9983361959457397)]\n",
      "[[[74.0, 1222.0], [172.0, 1222.0], [172.0, 1251.0], [74.0, 1251.0]], ('AMERI', 0.896643340587616)]\n",
      "[[[286.0, 1229.0], [464.0, 1229.0], [464.0, 1251.0], [286.0, 1251.0]], ('6.43% (of 311 reads)', 0.9685245752334595)]\n",
      "[[[495.0, 1231.0], [568.0, 1231.0], [568.0, 1250.0], [495.0, 1250.0]], ('normal', 0.9974880814552307)]\n",
      "[[[684.0, 1227.0], [747.0, 1227.0], [747.0, 1251.0], [684.0, 1251.0]], ('Tier 3', 0.9997832775115967)]\n",
      "[[[880.0, 1229.0], [1105.0, 1229.0], [1105.0, 1251.0], [880.0, 1251.0]], ('Uncertain Significance', 0.999556303024292)]\n",
      "[[[78.0, 1255.0], [175.0, 1255.0], [175.0, 1274.0], [78.0, 1274.0]], ('c.l646G>A', 0.9327661991119385)]\n",
      "[[[76.0, 1277.0], [163.0, 1277.0], [163.0, 1301.0], [76.0, 1301.0]], ('p.R549Q', 0.9999831318855286)]\n",
      "[[[74.0, 1303.0], [213.0, 1301.0], [213.0, 1325.0], [75.0, 1327.0]], ('g.6419164lC>T', 0.9326589107513428)]\n",
      "[[[78.0, 1349.0], [166.0, 1349.0], [166.0, 1377.0], [78.0, 1377.0]], ('CRLF2', 0.9999202489852905)]\n",
      "[[[286.0, 1353.0], [412.0, 1353.0], [412.0, 1377.0], [286.0, 1377.0]], ('43.0% (of 326', 0.9933373928070068)]\n",
      "[[[496.0, 1350.0], [536.0, 1355.0], [533.0, 1377.0], [493.0, 1373.0]], ('loss', 0.9821224212646484)]\n",
      "[[[686.0, 1353.0], [747.0, 1353.0], [747.0, 1377.0], [686.0, 1377.0]], ('Tier 3', 0.9978053569793701)]\n",
      "[[[880.0, 1354.0], [1103.0, 1354.0], [1103.0, 1377.0], [880.0, 1377.0]], ('Uncertain Significance', 0.9997192621231079)]\n",
      "[[[76.0, 1377.0], [154.0, 1377.0], [154.0, 1401.0], [76.0, 1401.0]], ('c.29C>T', 0.9999490976333618)]\n",
      "[[[282.0, 1378.0], [348.0, 1378.0], [348.0, 1402.0], [282.0, 1402.0]], ('reads)', 0.9994509816169739)]\n",
      "[[[76.0, 1402.0], [142.0, 1402.0], [142.0, 1427.0], [76.0, 1427.0]], ('p.P1OL', 0.9502907395362854)]\n",
      "[[[74.0, 1428.0], [209.0, 1426.0], [210.0, 1451.0], [75.0, 1452.0]], ('g.1202520G>A', 0.9808559417724609)]\n",
      "[[[77.0, 1469.0], [179.0, 1473.0], [178.0, 1499.0], [76.0, 1495.0]], ('EPCAM', 0.9996888041496277)]\n",
      "[[[284.0, 1475.0], [470.0, 1476.0], [469.0, 1501.0], [284.0, 1499.0]], ('36.0% (of 163 reads)', 0.9784929156303406)]\n",
      "[[[495.0, 1480.0], [568.0, 1480.0], [568.0, 1499.0], [495.0, 1499.0]], ('normal', 0.9971703886985779)]\n",
      "[[[686.0, 1476.0], [747.0, 1476.0], [747.0, 1500.0], [686.0, 1500.0]], ('Tier 3', 0.9970377087593079)]\n",
      "[[[880.0, 1478.0], [1103.0, 1478.0], [1103.0, 1500.0], [880.0, 1500.0]], ('Uncertain Significance', 0.9995231032371521)]\n",
      "[[[76.0, 1502.0], [158.0, 1502.0], [158.0, 1526.0], [76.0, 1526.0]], ('c.93C>G', 0.999785304069519)]\n"
     ]
    }
   ],
   "source": [
    "test_result = ocr.ocr(image_path, cls=True)\n",
    "for idx in range(len(test_result)):\n",
    "    res = test_result[idx]\n",
    "    for line in res:\n",
    "        print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "047cc6f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# boxes, texts, and probabilities from result\n",
    "res = test_result[0]\n",
    "boxes = [line[0] for line in res]\n",
    "texts = [line[1][0] for line in res]\n",
    "probabilities = [line[1][1] for line in res]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "21f7a285",
   "metadata": {},
   "outputs": [],
   "source": [
    "image_cv = cv2.imread(image_path)\n",
    "image_boxes = image_cv.copy()\n",
    "image_height, image_width = image_cv.shape[0], image_cv.shape[1]\n",
    "for box, text in zip(boxes, texts):\n",
    "    cv2.rectangle(image_boxes, (int(box[0][0]), int(box[0][1]), int(box[2][0]), int(box[2][1])), (0, 0, 255), 1)\n",
    "    cv2.putText(image_boxes, text, (int(box[0][0]), int(box[0][1])), cv2.FONT_HERSHEY_SCRIPT_SIMPLEX, 1, (222, 0, 0), 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b011260c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Returns the coordinates of the middle of the box\n",
    "def top_left(box):\n",
    "    return box[0]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "dd176be2",
   "metadata": {},
   "source": [
    "# Reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "037870ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "im = image_cv.copy()\n",
    "\n",
    "hor_boxes = []\n",
    "vert_boxes = []\n",
    "\n",
    "for box in boxes:\n",
    "    x_h, x_v = 0, int(box[0][0])\n",
    "    y_h, y_v = int(box[0][1]), 0 \n",
    "\n",
    "    width_h, width_v = image_width, int(box[2][0] - box[0][0])\n",
    "    height_h, height_v = int(box[2][1] - box[0][1]), image_height\n",
    "\n",
    "    hor_boxes.append([x_h, y_h, x_h + width_h, y_h + height_h])\n",
    "    vert_boxes.append([x_v, y_v, x_v + width_v, y_v + height_v])\n",
    "\n",
    "    cv2.rectangle(im, (x_h, y_h), (x_h + width_h, y_h + height_h), (255, 0, 0), 1)\n",
    "    cv2.rectangle(im, (x_v, y_v), (x_v + width_v, y_v + height_v), (0, 255, 0), 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b79f6cb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cv2.imwrite('hor_vert.jpg', im)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "63c78782",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.13.0\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "print(tf.__version__)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1279dd50",
   "metadata": {},
   "source": [
    "# Non Max Suppression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "512d758c",
   "metadata": {},
   "outputs": [],
   "source": [
    "hor_out = tf.image.non_max_suppression(\n",
    "    hor_boxes,\n",
    "    probabilities,\n",
    "    max_output_size = 1000,\n",
    "    iou_threshold = .1,\n",
    "    score_threshold = float('-inf'), \n",
    "    name = None\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "11b7325a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 0  2  3  4  6  7 14 15 16 17 18 19 24 27 29 30 32 33 36 38 40 41 43 44\n",
      " 46 49 51 52 53 54 55 56 57 60 66 68 69 70 71 76 78 79 80 85]\n"
     ]
    }
   ],
   "source": [
    "hor_lines = np.sort(np.array(hor_out))\n",
    "print(hor_lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f0804cea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "im_nms = image_cv.copy()\n",
    "for val in hor_lines:\n",
    "    cv2.rectangle(im_nms, (int(hor_boxes[val][0]), int(hor_boxes[val][1])), (int(hor_boxes[val][2]), int(hor_boxes[val][3])), (0, 0, 255), 1)\n",
    "cv2.imwrite('im_nms.jpg', im_nms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "53f03d59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 7  8 49 51 60 69]\n"
     ]
    }
   ],
   "source": [
    "vert_out = tf.image.non_max_suppression(\n",
    "    vert_boxes,\n",
    "    probabilities,\n",
    "    max_output_size = 1000,\n",
    "    iou_threshold = .1,\n",
    "    score_threshold = float('-inf'), \n",
    "    name = None\n",
    ")\n",
    "vert_lines = np.sort(np.array(vert_out))\n",
    "print(vert_lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "638ed581",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "im_nms2 = im_nms.copy()\n",
    "\n",
    "for val in vert_lines:\n",
    "    cv2.rectangle(im_nms2, (int(vert_boxes[val][0]), int(vert_boxes[val][1])), (int(vert_boxes[val][2]), int(vert_boxes[val][3])), (255, 0, 0), 1)\n",
    "cv2.imwrite('im_nms2.jpg', im_nms2) "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "29dd6287",
   "metadata": {},
   "source": [
    "# Change to CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c6a1c475",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(44, 6)\n"
     ]
    }
   ],
   "source": [
    "out_array = [[\"\" for i in range(len(vert_lines))] for j in range(len(hor_lines))]\n",
    "print(np.array(out_array).shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "12c09780",
   "metadata": {},
   "outputs": [],
   "source": [
    "underordered_boxes = []\n",
    "for i in vert_lines:\n",
    "    underordered_boxes.append(vert_boxes[i][0])\n",
    "ordered_boxes = np.array(np.argsort(underordered_boxes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1a43a198",
   "metadata": {},
   "outputs": [],
   "source": [
    "# intersection over union\n",
    "def iou(box_1, box_2):\n",
    "\n",
    "    x_1 = max(box_1[0], box_2[0])\n",
    "    y_1 = max(box_1[1], box_2[1])\n",
    "    x_2 = min(box_1[2], box_1[2])\n",
    "    y_2 = min(box_1[3], box_2[3])\n",
    "\n",
    "    inter = abs(max(x_2 - x_1, 0) * max(y_2 - y_1, 0))\n",
    "    if inter == 0:\n",
    "        return 0\n",
    "\n",
    "    area_1 = abs((box_1[2] - box_1[0]) * (box_1[3] - box_1[1]))\n",
    "    area_2 = abs((box_2[2] - box_2[0]) * (box_2[3] - box_2[1]))\n",
    "\n",
    "    return inter / float(area_1 + area_2 - inter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a169da2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain the intersection of a given pair of vertical and horizontal boxes\n",
    "def intersection(box_1, box_2):\n",
    "    return [box_2[0], box_1[1], box_2[2], box_1[3]]\n",
    "\n",
    "for i in range(len(hor_lines)):\n",
    "    for j in range(len(vert_lines)):\n",
    "        resultant = intersection(hor_boxes[hor_lines[i]], vert_boxes[vert_lines[ordered_boxes[j]]])\n",
    "        for b in range(len(boxes)):\n",
    "            the_box = [boxes[b][0][0], boxes[b][0][1], boxes[b][2][0], boxes[b][2][1]]\n",
    "            if iou(resultant, the_box) > .1:\n",
    "                out_array[i][j] = texts[b]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c6bc9f66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Associated With</td>\n",
       "      <td>Clinical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Approved Therapies</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Classification</td>\n",
       "      <td>Gene / Variant</td>\n",
       "      <td>Allelic Fraction</td>\n",
       "      <td>Function</td>\n",
       "      <td>Assessment</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Other Indications</td>\n",
       "      <td>Gene /Variant</td>\n",
       "      <td>Allelic Fraction</td>\n",
       "      <td>Gastric Cancer</td>\n",
       "      <td>Resistance</td>\n",
       "      <td>Trials</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tier 3</td>\n",
       "      <td>AMERICRLF2EPCAM</td>\n",
       "      <td>6.43% (of 311 reads)43.0% (of 32636.0% (of 163...</td>\n",
       "      <td>normallossnormal</td>\n",
       "      <td>Uncertain SignificanceUncertain SignificanceUn...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>afatinib</td>\n",
       "      <td>ERBB3</td>\n",
       "      <td>48.0% (of 629</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>dacomitinib</td>\n",
       "      <td>g.56101239G&gt;A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>emtansine</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>erdafitinib</td>\n",
       "      <td>FGFR4</td>\n",
       "      <td>48.0% (of 145 reads)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>futibatinib</td>\n",
       "      <td>c.1871A&gt;C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>infigratinib</td>\n",
       "      <td>g.177096106A&gt;C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>lenvatinib</td>\n",
       "      <td>Tier 3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>neratinib</td>\n",
       "      <td>Tier 3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>osimertinib</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pertuzumab</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ponatinib</td>\n",
       "      <td>UncertainSignificance</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>regorafenib</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ruxolitinib</td>\n",
       "      <td>JAK3</td>\n",
       "      <td>58.0% (of 262</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>trastuzumab</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tucatinib</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     3                      0  \\\n",
       "0                                               \n",
       "1   Approved Therapies                          \n",
       "2       Classification         Gene / Variant   \n",
       "3    Other Indications          Gene /Variant   \n",
       "4               Tier 3        AMERICRLF2EPCAM   \n",
       "5             afatinib                  ERBB3   \n",
       "6          dacomitinib          g.56101239G>A   \n",
       "7            emtansine                          \n",
       "8          erdafitinib                  FGFR4   \n",
       "9          futibatinib              c.1871A>C   \n",
       "10        infigratinib         g.177096106A>C   \n",
       "11          lenvatinib                 Tier 3   \n",
       "12           neratinib                 Tier 3   \n",
       "13         osimertinib              Uncertain   \n",
       "14          pertuzumab                          \n",
       "15           ponatinib  UncertainSignificance   \n",
       "16         regorafenib                          \n",
       "17         ruxolitinib                   JAK3   \n",
       "18         trastuzumab                          \n",
       "19           tucatinib                          \n",
       "\n",
       "                                                    1                 2  \\\n",
       "0                                                                         \n",
       "1                                                                         \n",
       "2                                    Allelic Fraction          Function   \n",
       "3                                    Allelic Fraction    Gastric Cancer   \n",
       "4   6.43% (of 311 reads)43.0% (of 32636.0% (of 163...  normallossnormal   \n",
       "5                                       48.0% (of 629                 0   \n",
       "6                                                   0                 0   \n",
       "7                                                                         \n",
       "8                                48.0% (of 145 reads)                 0   \n",
       "9                                                   0                 0   \n",
       "10                                                  0                 0   \n",
       "11                                                  0                 0   \n",
       "12                                                  0                 0   \n",
       "13                                                  0                 0   \n",
       "14                                                                        \n",
       "15                                                  0                 0   \n",
       "16                                                                        \n",
       "17                                      58.0% (of 262                 0   \n",
       "18                                                                        \n",
       "19                                                                        \n",
       "\n",
       "                                                    4         5  \n",
       "0                                     Associated With  Clinical  \n",
       "1                                                   0         0  \n",
       "2                                          Assessment         0  \n",
       "3                                          Resistance    Trials  \n",
       "4   Uncertain SignificanceUncertain SignificanceUn...         0  \n",
       "5                                                   0         0  \n",
       "6                                                   0         0  \n",
       "7                                                   0         0  \n",
       "8                                                   0         0  \n",
       "9                                                   0         0  \n",
       "10                                                  0         0  \n",
       "11                                                  0         0  \n",
       "12                                                  0         0  \n",
       "13                                                  0         0  \n",
       "14                                                  0         0  \n",
       "15                                                  0         0  \n",
       "16                                                  0         0  \n",
       "17                                                  0         0  \n",
       "18                                                  0         0  \n",
       "19                                                  0         0  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(out_array)\n",
    "\n",
    "# remove duplicates entries in rows and replace with 0\n",
    "# replace zero entries with NaN \n",
    "mask = df.apply(pd.Series.duplicated, 1) & df.astype(bool)\n",
    "removed_df = df.mask(mask, 0)\n",
    "for col in removed_df.columns:\n",
    "    removed_df[col].replace(to_replace = 0, value = pd.NA, inplace=True)\n",
    "removed_df.groupby([3], axis=0, as_index=False).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0d7b547d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('sample.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ee0ac82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "removed_df.to_csv('fixed_sample.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d9672ded",
   "metadata": {},
   "outputs": [],
   "source": [
    "# still need to fix the problem with leftmost column not being grouped together as one entry\n",
    "# figure out how to deal with duplicate entries due to empty spaces on the document"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "612e377e",
   "metadata": {},
   "source": [
    "### Grouping Correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "eb7f51c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PIL import Image\n",
    "\n",
    "rgb_image = Image.open(image_path)\n",
    "pix = rgb_image.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "715de697",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(255, 255, 255)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pix[tuple(top_left(boxes[0]))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "12f076c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "color_by_row = ['' for i in range(len(hor_lines))]\n",
    "len(color_by_row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "34bb7c9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import webcolors\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a8c20d1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'white',\n",
       " 'gray',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'white',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'gray',\n",
       " 'white',\n",
       " 'white']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hor_color_indices = [hor_boxes[line][1] for line in hor_lines]\n",
    "hor_color_indices\n",
    "\n",
    "for i in range(len(color_by_row)):\n",
    "    color = pix[(50, hor_color_indices[i])]\n",
    "    try:\n",
    "        color_by_row[i] = webcolors.rgb_to_name(color)\n",
    "    except ValueError:\n",
    "        color_by_row[i] = 'gray'\n",
    "color_by_row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c69efa1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>Accession ID: NGSO-221209001</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Page 5 of 76</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>Patient Name: LAN ZHENG,MEI-CHAN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Diagnosis: gastric cancer</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Report Date: Dec 19,2022</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YOUR LAB</td>\n",
       "      <td>Report Date: Dec 19,2022</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Approved Therapies</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Associated With</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Gene /Variant</td>\n",
       "      <td>Allelic Fraction</td>\n",
       "      <td>Gastric Cancer</td>\n",
       "      <td>Other Indications</td>\n",
       "      <td>Resistance</td>\n",
       "      <td>Trials</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>pertuzumab</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>trastuzumab</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>emtansine</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>tucatinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ERBB3</td>\n",
       "      <td>48.0% (of 629</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>afatinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p.Rl127H</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>g.56101239G&gt;A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>dacomitinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Tier 3</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>neratinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Uncertain</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>osimertinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Significance</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FGFR4</td>\n",
       "      <td>48.0% (of 145 reads)</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>erdafitinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>c.1871A&gt;C</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>futibatinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>p.N624T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>g.177096106A&gt;C</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>infigratinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Tier 3</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>lenvatinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Uncertain</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ponatinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Significance</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ponatinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regorafenib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>JAK3</td>\n",
       "      <td>58.0% (of 262</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ruxolitinib</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>c.2660G&gt;A</td>\n",
       "      <td>reads)</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>p.R887H</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>g.17832539C&gt;T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Tier 3</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Uncertain</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Significance</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Variants of Unknown Clinical Significance</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Gene / Variant</td>\n",
       "      <td>Allelic Fraction</td>\n",
       "      <td>Function</td>\n",
       "      <td>Classification</td>\n",
       "      <td>Assessment</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>AMERI</td>\n",
       "      <td>6.43% (of 311 reads)</td>\n",
       "      <td>normal</td>\n",
       "      <td>Tier 3</td>\n",
       "      <td>Uncertain Significance</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>c.l646G&gt;A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>p.R549Q</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>g.6419164lC&gt;T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>CRLF2</td>\n",
       "      <td>43.0% (of 326</td>\n",
       "      <td>loss</td>\n",
       "      <td>Tier 3</td>\n",
       "      <td>Uncertain Significance</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>c.29C&gt;T</td>\n",
       "      <td>reads)</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>p.P1OL</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>g.1202520G&gt;A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>gray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>EPCAM</td>\n",
       "      <td>36.0% (of 163 reads)</td>\n",
       "      <td>normal</td>\n",
       "      <td>Tier 3</td>\n",
       "      <td>Uncertain Significance</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>c.93C&gt;G</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            0  \\\n",
       "0                                               \n",
       "1                                               \n",
       "2                                               \n",
       "3                                    YOUR LAB   \n",
       "4                                               \n",
       "5                                               \n",
       "6                               Gene /Variant   \n",
       "7                                               \n",
       "8                                               \n",
       "9                                               \n",
       "10                                              \n",
       "11                                      ERBB3   \n",
       "12                                   p.Rl127H   \n",
       "13                              g.56101239G>A   \n",
       "14                                     Tier 3   \n",
       "15                                  Uncertain   \n",
       "16                               Significance   \n",
       "17                                      FGFR4   \n",
       "18                                  c.1871A>C   \n",
       "19                                    p.N624T   \n",
       "20                             g.177096106A>C   \n",
       "21                                     Tier 3   \n",
       "22                                  Uncertain   \n",
       "23                               Significance   \n",
       "24                                              \n",
       "25                                       JAK3   \n",
       "26                                  c.2660G>A   \n",
       "27                                    p.R887H   \n",
       "28                              g.17832539C>T   \n",
       "29                                     Tier 3   \n",
       "30                                  Uncertain   \n",
       "31                               Significance   \n",
       "32  Variants of Unknown Clinical Significance   \n",
       "33                             Gene / Variant   \n",
       "34                                      AMERI   \n",
       "35                                  c.l646G>A   \n",
       "36                                    p.R549Q   \n",
       "37                              g.6419164lC>T   \n",
       "38                                      CRLF2   \n",
       "39                                    c.29C>T   \n",
       "40                                     p.P1OL   \n",
       "41                               g.1202520G>A   \n",
       "42                                      EPCAM   \n",
       "43                                    c.93C>G   \n",
       "\n",
       "                                   1                          2  \\\n",
       "0       Accession ID: NGSO-221209001                       <NA>   \n",
       "1   Patient Name: LAN ZHENG,MEI-CHAN                       <NA>   \n",
       "2                                     Diagnosis: gastric cancer   \n",
       "3           Report Date: Dec 19,2022                       <NA>   \n",
       "4                                                                 \n",
       "5                                                                 \n",
       "6                   Allelic Fraction             Gastric Cancer   \n",
       "7                                                                 \n",
       "8                                                                 \n",
       "9                                                                 \n",
       "10                                                                \n",
       "11                     48.0% (of 629                       <NA>   \n",
       "12                              <NA>                       <NA>   \n",
       "13                              <NA>                       <NA>   \n",
       "14                              <NA>                       <NA>   \n",
       "15                              <NA>                       <NA>   \n",
       "16                              <NA>                       <NA>   \n",
       "17              48.0% (of 145 reads)                       <NA>   \n",
       "18                              <NA>                       <NA>   \n",
       "19                              <NA>                       <NA>   \n",
       "20                              <NA>                       <NA>   \n",
       "21                              <NA>                       <NA>   \n",
       "22                              <NA>                       <NA>   \n",
       "23                              <NA>                       <NA>   \n",
       "24                                                                \n",
       "25                     58.0% (of 262                       <NA>   \n",
       "26                            reads)                       <NA>   \n",
       "27                              <NA>                       <NA>   \n",
       "28                              <NA>                       <NA>   \n",
       "29                              <NA>                       <NA>   \n",
       "30                              <NA>                       <NA>   \n",
       "31                              <NA>                       <NA>   \n",
       "32                              <NA>                       <NA>   \n",
       "33                  Allelic Fraction                   Function   \n",
       "34              6.43% (of 311 reads)                     normal   \n",
       "35                              <NA>                       <NA>   \n",
       "36                              <NA>                       <NA>   \n",
       "37                              <NA>                       <NA>   \n",
       "38                     43.0% (of 326                       loss   \n",
       "39                            reads)                       <NA>   \n",
       "40                              <NA>                       <NA>   \n",
       "41                              <NA>                       <NA>   \n",
       "42              36.0% (of 163 reads)                     normal   \n",
       "43                              <NA>                       <NA>   \n",
       "\n",
       "                     3                         4             5  color  \n",
       "0                 <NA>                      <NA>  Page 5 of 76  white  \n",
       "1                 <NA>                      <NA>          <NA>  white  \n",
       "2                 <NA>  Report Date: Dec 19,2022          <NA>  white  \n",
       "3                 <NA>                      <NA>          <NA>  white  \n",
       "4   Approved Therapies                      <NA>          <NA>  white  \n",
       "5                                Associated With      Clinical   gray  \n",
       "6    Other Indications                Resistance        Trials   gray  \n",
       "7           pertuzumab                      <NA>          <NA>  white  \n",
       "8          trastuzumab                      <NA>          <NA>  white  \n",
       "9            emtansine                      <NA>          <NA>  white  \n",
       "10           tucatinib                      <NA>          <NA>  white  \n",
       "11            afatinib                      <NA>          <NA>   gray  \n",
       "12                <NA>                      <NA>          <NA>   gray  \n",
       "13         dacomitinib                      <NA>          <NA>   gray  \n",
       "14           neratinib                      <NA>          <NA>   gray  \n",
       "15         osimertinib                      <NA>          <NA>   gray  \n",
       "16                <NA>                      <NA>          <NA>   gray  \n",
       "17         erdafitinib                      <NA>          <NA>  white  \n",
       "18         futibatinib                      <NA>          <NA>  white  \n",
       "19                <NA>                      <NA>          <NA>  white  \n",
       "20        infigratinib                      <NA>          <NA>  white  \n",
       "21          lenvatinib                      <NA>          <NA>  white  \n",
       "22           ponatinib                      <NA>          <NA>  white  \n",
       "23           ponatinib                      <NA>          <NA>  white  \n",
       "24         regorafenib                      <NA>          <NA>  white  \n",
       "25         ruxolitinib                      <NA>          <NA>   gray  \n",
       "26                <NA>                      <NA>          <NA>   gray  \n",
       "27                <NA>                      <NA>          <NA>   gray  \n",
       "28                <NA>                      <NA>          <NA>   gray  \n",
       "29                <NA>                      <NA>          <NA>   gray  \n",
       "30                <NA>                      <NA>          <NA>   gray  \n",
       "31                <NA>                      <NA>          <NA>   gray  \n",
       "32                <NA>                      <NA>          <NA>  white  \n",
       "33      Classification                Assessment          <NA>   gray  \n",
       "34              Tier 3    Uncertain Significance          <NA>  white  \n",
       "35                <NA>                      <NA>          <NA>  white  \n",
       "36                <NA>                      <NA>          <NA>  white  \n",
       "37                <NA>                      <NA>          <NA>  white  \n",
       "38              Tier 3    Uncertain Significance          <NA>   gray  \n",
       "39                <NA>                      <NA>          <NA>   gray  \n",
       "40                <NA>                      <NA>          <NA>   gray  \n",
       "41                <NA>                      <NA>          <NA>   gray  \n",
       "42              Tier 3    Uncertain Significance          <NA>  white  \n",
       "43                <NA>                      <NA>          <NA>  white  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "removed_df['color'] = color_by_row\n",
    "removed_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "43fc3cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_rows(df):\n",
    "    # Create an empty list to store rows of the new DataFrame\n",
    "    new_rows = []\n",
    "\n",
    "    current_color = None\n",
    "    current_row = None\n",
    "\n",
    "    for index, row in df.iterrows():\n",
    "        if row['color'] == current_color:\n",
    "            current_row = [str(current_row[i]) + '\\n' + str(row[i]) if str(row[i]) != '<NA>' else current_row[i] for i in range(len(row)-1)]\n",
    "        else:\n",
    "            # If the color of the current row is different or 'color' column does not exist,\n",
    "            # create a new row in the new DataFrame and update the current color group\n",
    "            if current_row is not None:\n",
    "                new_rows.append(current_row)\n",
    "            current_row = row.tolist()\n",
    "            current_color = row['color']\n",
    "\n",
    "    # Append the last row to the new DataFrame\n",
    "    if current_row is not None:\n",
    "        new_rows.append(current_row)\n",
    "\n",
    "    # Convert the list of rows into a new DataFrame\n",
    "    new_df = pd.DataFrame(new_rows, columns=df.columns)\n",
    "    return new_df\n",
    "\n",
    "# Assuming you have already defined the 'group_rows' function and provided the DataFrame 'removed_df' as an argument\n",
    "result_df = group_rows(removed_df)\n",
    "result_df = result_df.drop('color', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "00c751af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_value(value):\n",
    "    return str(value).replace('<NA>', '').strip()\n",
    "for index, row in result_df.iterrows():\n",
    "    result_df.loc[index] = [clean_value(value) for value in row]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ec2d0bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df.to_csv('groupedsample.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dde8248",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  },
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
